app/assets/javascripts/visualizations/highvis/binaryTree.coffee

```diff
@@ -0,0 +1,393 @@
+              @left = new BinaryTree(@)
+            @left.data = data
+
+          else
+            console.log "Error inserting #{data} into left child of tree."
+        else if pos is 'right'
+          if @data isnt null
+            if @right is null
+              @right = new BinaryTree(@)
+            @right.data = data
+          else
+            console.log "Error inserting #{data} into right child of tree."
+        else
+          @data = data
+
+      # Delete a single datum in the tree at @data,
+      # or the @data member of the tree located at
+      # pos = 'left' or pos = 'right'
+      ###
+      # WARNING:  MUTATES THE BINARY TREE
+      ###
+      deleteData: (pos = null) ->
+        if pos is 'right'
+          if @right.is_terminal()
+            @right = null
+          else
+            console.log "Error deleting #{@right.data}, results in invalid binary tree."
+            null
+        else if pos is 'left'
+          if @left.is_terminal()
+            @left = null
+          else
+            console.log "Error deleting #{@left.data}, results in invalid binary tree."
+            null
+        else
+          if @is_terminal()
+            @data = null
+          else
+            console.log "Error deleting #{@data}, results in invalid binary tree."
+            null
+
+      # Allows the user to index into the tree, following Preorder traversal:
+      # ROOT, left, right
+      index: (index) ->
+        @__access(index)
+
+      # Determine how far away the node at position 'index' is from the root
+      # of the binary tree.  This is used to determine how long the mutation
+      # tree at a given point can be to maintain the maximum depth of the
+      # tree during the point mutation genetic operation.
+      depthAtPoint: (index, curDepth = 1) ->
+        @__access(index, false, curDepth)
+      
+      # Internal method used to abstract the index and depth at point member
+      # functions.
+      ###
+      # WARNING:  INTERNAL METHOD.  DO NOT CALL
+      ###
+      __access: (index, value = true, curDepth = 1) ->
+        if index is 0
+          return if value is true then @ else curDepth
+        leftSize = if @left is null then 0 else @left.treeSize()
+        rightSize = if @right is null then 0 else @right.treeSize()
+        if index > leftSize + rightSize
+          -1
+        else if index > leftSize
+          if @right isnt null then @right.__access(index - leftSize - 1, value, curDepth + 1) else -1
+        else
+          if @left isnt null then @left.__access(index - 1, value, curDepth + 1) else -1
+      
+      # Given a tree, replace it with a randomly-generated tree whose maximum
+      # depth is given by maxDepth.
+      ###
+      # WARNING:  MUTATES THE BINARY TREE
+      ###
+      generate: (maxDepth = 10, curDepth = 1) ->
+        if curDepth is maxDepth
+          @insertData(window.BinaryTree.terminals[Math.floor(Math.random() * window.BinaryTree.terminals.length)])
```

Comment:
you don't need 'window' -> the default scope is window, so you already have access to it